### PR TITLE
Implement per stream timeout

### PR DESCRIPTION
**Pull request by pcalley**
Allows `timeout` parameter to be optionally added under `origin_server` in the bootstrap config in order to add a per stream timeout for each connection to the upstream server. 

Timeout is a duration in milliseconds and defaults to no timeout. 

Jitter of timeout values to come in subsequent PR.

### CODE REVIEW

File: internal/app/upstream/client_test.go [hunk 1]
```diff
@@ -358,6 +362,62 @@ func TestOpenStreamShouldRetryIfSendFails(t *testing.T) {
...
+		var val int64 = 0
+		if scope.Snapshot().Counters() != nil &&
+			scope.Snapshot().Counters()["mock.lds.stream_retry+key=aggregated_key"] != nil {
+			val = scope.Snapshot().Counters()["mock.lds.stream_retry+key=aggregated_key"].Value()
+		}
+		if val > 0 {
+			break
+		}
+	}
+	done()
```
Comment (samrabelachew) [inline internal/app/upstream/client_test.go:389]: One question for my understanding: why do we call `defer done()` if we run `done()` here?

**Review by samrabelachew — CHANGES_REQUESTED** [review]

File: api/protos/bootstrap/v1/bootstrap.proto [hunk 2]
```diff
@@ -47,6 +47,10 @@ message Upstream {
     // From https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/Duration
     // Duration only lets us represent time in 's'
     string keep_alive_time = 2;
+
+    // Timeout in milliseconds for upstream connection stream.
+    // If unset defaults to no timeout.
+    int64 timeout = 3;
```
Comment (jessicayuen) [inline api/protos/bootstrap/v1/bootstrap.proto:53]: Can you explain how this differentiates from the the `keep_alive_time` above?

Comment (pcalley) [reply api/protos/bootstrap/v1/bootstrap.proto:53]: The `keep_alive_time` tells the stream to send a ping (defaults every 5m) and would close any broken streams. The `timeout` cancels the context that each stream uses on a fixed interval forcing the connection to close and reopen. My assumption is that when the stream is cancelled it forces parts of the cache to be refreshed.

File: internal/app/upstream/client_test.go (hunk 1 above)
Comment (pcalley) [reply internal/app/upstream/client_test.go]: Good catch I think the second done is not necessary.

File: api/protos/bootstrap/v1/bootstrap.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (pcalley) [reply api/protos/bootstrap/v1/bootstrap.proto:53]: Fixed.

File: internal/app/upstream/client_test.go [hunk 3]
```diff
@@ -358,6 +362,60 @@ func TestOpenStreamShouldRetryIfSendFails(t *testing.T) {
...
+	for {
+		var val int64 = 0
+		if scope.Snapshot().Counters() != nil &&
+			scope.Snapshot().Counters()["mock.lds.stream_retry+key=aggregated_key"] != nil {
+			val = scope.Snapshot().Counters()["mock.lds.stream_retry+key=aggregated_key"].Value()
+		}
+		if val > 0 {
+			break
+		}
+	}
```
Comment (jessicayuen) [inline internal/app/upstream/client_test.go:388]: If the test condition fails, this loop will cause the test to hang indefinitely is that correct? Can we make this test automatically fail if a set duration has passed?

**Review by jessicayuen — COMMENTED** [review]
LGTM to me, just the one outstanding comment.

File: internal/app/upstream/client_test.go (hunk 3 above)
Comment (pcalley) [reply internal/app/upstream/client_test.go]: Yes this is correct. The latest commit should check for this condition and fail the test after 5s if the success condition isn't met.

**Review by jessicayuen — APPROVED** [review]

**Review by samrabelachew — APPROVED** [review]

**Pull request merged**